mlir/test/Dialect/Vector/canonicalize.mlir

```diff
@@ -782,35 +782,47 @@ func.func @fold_extract_shapecast(%arg0 : vector<5x1x3x2xf32>,
 
 // -----
 
-// CHECK-LABEL: fold_extract_shapecast_negative
-//       CHECK:   %[[V:.*]] = vector.shape_cast %{{.*}} : vector<16xf32> to vector<2x4x2xf32>
-//       CHECK:   %[[R:.*]] = vector.extract %[[V]][1] : vector<4x2xf32> from vector<2x4x2xf32>
-//       CHECK:   return %[[R]] : vector<4x2xf32>
-func.func @fold_extract_shapecast_negative(%arg0 : vector<16xf32>) -> vector<4x2xf32> {
-  %0 = vector.shape_cast %arg0 : vector<16xf32> to vector<2x4x2xf32>
-  %r = vector.extract %0[1] : vector<4x2xf32> from vector<2x4x2xf32>
-  return %r : vector<4x2xf32>
+// CHECK-LABEL: fold_extract_shapecast_0d_result
+//  CHECK-SAME: %[[IN:.*]]: vector<1x1x1xf32>
+//       CHECK:   %[[R:.*]] = vector.extract %[[IN]][0, 0, 0] : f32 from vector<1x1x1xf32>
+//       CHECK:   return %[[R]] : f32
+func.func @fold_extract_shapecast_0d_result(%arg0 : vector<1x1x1xf32>) -> f32 {
+  %0 = vector.shape_cast %arg0 : vector<1x1x1xf32> to vector<f32>
+  %r = vector.extract %0[] : f32 from vector<f32>
+  return %r : f32
 }
 
 // -----
 
-// CHECK-LABEL: dont_fold_0d_extract_shapecast
-//       CHECK:   %[[V:.*]] = vector.shape_cast %{{.*}} : vector<f32> to vector<1xf32>
-//       CHECK:   %[[R:.*]] = vector.extract %[[V]][0] : f32 from vector<1xf32>
+// CHECK-LABEL: fold_extract_shapecast_0d_source
+//  CHECK-SAME: %[[IN:.*]]: vector<f32>
+//       CHECK:   %[[R:.*]] = vector.extract %[[IN]][] : f32 from vector<f32>
 //       CHECK:   return %[[R]] : f32
-func.func @dont_fold_0d_extract_shapecast(%arg0 : vector<f32>) -> f32 {
+func.func @fold_extract_shapecast_0d_source(%arg0 : vector<f32>) -> f32 {
   %0 = vector.shape_cast %arg0 : vector<f32> to vector<1xf32>
   %r = vector.extract %0[0] : f32 from vector<1xf32>
   return %r : f32
 }
 
 // -----
 
-// CHECK-LABEL: fold_extract_shapecast_to_shapecast
+// CHECK-LABEL: fold_extract_shapecast_negative
+//       CHECK:   %[[V:.*]] = vector.shape_cast %{{.*}} : vector<16xf32> to vector<2x4x2xf32>
+//       CHECK:   %[[R:.*]] = vector.extract %[[V]][1] : vector<4x2xf32> from vector<2x4x2xf32>
+//       CHECK:   return %[[R]] : vector<4x2xf32>
+func.func @fold_extract_shapecast_negative(%arg0 : vector<16xf32>) -> vector<4x2xf32> {
+  %0 = vector.shape_cast %arg0 : vector<16xf32> to vector<2x4x2xf32>
+  %r = vector.extract %0[1] : vector<4x2xf32> from vector<2x4x2xf32>
+  return %r : vector<4x2xf32>
+}
+
+// -----
+
+// CHECK-LABEL: canonicalize_extract_shapecast_to_shapecast
 //  CHECK-SAME: (%[[ARG:.+]]: vector<3x4xf32>)
 //       CHECK:   %[[R:.+]] = vector.shape_cast %[[ARG]] : vector<3x4xf32> to vector<12xf32>
 //       CHECK:   return %[[R]]
-func.func @fold_extract_shapecast_to_shapecast(%arg0 : vector<3x4xf32>) -> vector<12xf32> {
```

Comment:
Why change the name to canonicalize? This brakes consistency with other tests.

Reply:
It's to reflect that this test is checking a canonicalization pattern. There are folders for extractop + shapecast as well as canonicalizers. With this name, i can tell that this is testing the canonicalization pattern instead.

I can change it back, but I think naming it "canonicalize" tells me what exactly it's testing.

Reply:
IIUC, the only thing that makes this a "canonicalization" is including the corresponding pattern, [foldExtractFromShapeCastToShapeCast](https://github.com/llvm/llvm-project/blob/4086ead63c7e0b56b3b07873117bb7ad7a02d41c/mlir/lib/Dialect/Vector/IR/VectorOps.cpp#L2190-L2209), under [ExtractOp::getCanonicalizationPatterns](https://github.com/llvm/llvm-project/blob/4086ead63c7e0b56b3b07873117bb7ad7a02d41c/mlir/lib/Dialect/Vector/IR/VectorOps.cpp#L2259-L2265). 

Given that this file runs with `-canonicalize`, I presume that all tested patterns are included under some `getCanonicalizationPattern` method? So there's nothing special about this pattern, right? 

For consistency (with e.g. the existing tests and the pattern name), I'd keep it as is.

Reply:
> Given that this file runs with -canonicalize, I presume that all tested patterns are included under some getCanonicalizationPattern method? So there's nothing special about this pattern, right?

The folding happening here is a side effect of the canonicalizer using greedy pattern rewriter, which folds first and then applies a pattern. It is special because folding is not a pattern.

Regarding consistency, there are other patterns like "canonicalize_broadcast_shapecast_to_broadcast" and "fold_broadcast_shapecast" which differ based on what happens.

But, for this PR I'll rename it back to fold_, we can discuss later what we should use for this, I don't really mind :)